src/libtiled/tilesetmanager.cpp

```diff
@@ -57,10 +57,7 @@ TilesetManager::TilesetManager():
 }
 
 TilesetManager::~TilesetManager()
-{
-    // Assert that there are no remaining tileset instances
-    Q_ASSERT(mTilesets.isEmpty());
```

Comment:
We can't just remove this assertion. It's there because the destructor of `Tileset` will attempt to remove the tileset from the tileset manager, re-creating a TilesetManager if it was destroyed before that:

https://github.com/bjorn/tiled/blob/752df93430d2413a86fbb2fb04d726c80e030fa3/src/libtiled/tileset.cpp#L76-L78

Which would then run into this assertion instead

https://github.com/bjorn/tiled/blob/752df93430d2413a86fbb2fb04d726c80e030fa3/src/libtiled/tilesetmanager.cpp#L129-L131

In addition, the assertions simply try to make sure there are no `Tileset` instances "leaking" somewhere. Would it be a problem to clear the cache before destroying the tileset manager? This happens here:

https://github.com/bjorn/tiled/blob/752df93430d2413a86fbb2fb04d726c80e030fa3/src/tiled/mainwindow.cpp#L758

Reply:
Ah, I see. I'll try to work on avoiding the leaked resources then. I didn't really consider the leaking issue, because it's only leaking when the app closes.

Reply:
I've pushed a new version that purges all loaded submap instances at the start of the tileset manager destructor. This ensures that all tileset destructors run before the tileset manager's destructor finishes.